### ISSUE TITLE
Specify whether ESNI key is provided by hidden or by fronting server

### ISSUE THREAD

**Issue by huitema** [ekr/draft-rescorla-tls-esni#19]
The text in section 4, Publishing the SNI Encryption Key, specifies that _if a client queries example.com, the ESNI TXT name is _esni.example.com_. That's fine, but suppose that the client queries "hidden.example.com". The client needs to learn that "hidden.example.com" is fronted by "fronting.example.com" -- per PR #18, the client will do that by obtaining an ALT-SVC record. Is the client suppose to get the ESNI key from "_esni.hidden.example.com", or from "_esni.fronting.example.com" ?

**Comment by chris-wood** [comment]
I agree with @martinthomson here. I didn't see the client querying fronting.example.com.

**Comment by huitema** [comment]
That means all servers that share the same IP need to publish the same key. Are we confident that there is no synch issue during key rollover?

**Comment by chris-wood** [comment]
Servers can publish multiple ESNI keys with different labels in a record. That way they can retire one key and start using another as needed. Though it might be good (necessary?) to require clients to use one of those keys.

**Comment by kazuho** [comment]
> Servers can publish multiple ESNI keys with different labels in a record. That way they can retire one key and start using another as needed.

While that is possible, I am not sure if it would be the approach that would generally be used.

The issue is identical to the encryption key of the session tickets. In case of session tickets, we encrypt the ticket using one key, and send the identifier of the key along with the encrypted ticket.

When we rollover the key, we first deploy the new key to the server as well as keeping the old one alive for decryption only. And we retire the old key after certain amount of period that we become sure that it is no longer used.

I'd assume that we can and will use the same approach for ESNIKeys.

**Comment by kazuho** [comment]
PS. therefore, it is possible to have just one label for the set of keys rather than having labels for each of the keys. In that approach, a server will identify the key being used by looking at the `(label, group)` tuple.

The benefit will be that you will save the space of N-1 labels assuming that you are sending N key shares. Identification of the key on the server-side might seem complicated, but I actually think that it simplifies things, because a server operator is anyways expected to deploy the entire set of keys found in ESNIKeys at the same moment.

**Comment by kazuho** [comment]
> therefore, it is possible to have just one label for the set of keys rather than having labels for each of the keys (assuming that we can remove the "choose one at random" rule)

To me it seems that this is necessary to prevent downgrade attack. Hence #24.

**Comment by huitema** [comment]
How do we prevent labels from becoming identifiers?

**Comment by chris-wood** [comment]
Labels as identifiers seem problematic in both approaches -- current and #24.

**Comment by chris-wood** [comment]
Now that #43 is merged, which removes the label altogether, and we've added text describing the anonymity set for hidden servers, I think this issue is covered. @huitema do you agree?
